gradoop-data-integration/src/main/java/org/gradoop/dataintegration/transformation/ConnectNeighbors.java

```diff
@@ -0,0 +1,150 @@
+import org.gradoop.dataintegration.transformation.functions.Neighborhood;
+import org.gradoop.flink.model.api.operators.UnaryGraphToGraphOperator;
+import org.gradoop.flink.model.impl.epgm.LogicalGraph;
+import org.gradoop.flink.model.impl.operators.subgraph.functions.LabelIsIn;
+
+import java.util.List;
+
+/**
+ * This graph transformation adds new edges to the graph. Those edges are created if a vertex of a
+ * user-defined label has two neighbors of another user-defined label. A bidirectional (two edges
+ * in gradoop) edge is then created between those two neighbors.
+ */
+public class ConnectNeighbors implements UnaryGraphToGraphOperator {
+
+  /**
+   * The label of the vertices the neighborhood is connected.
+   */
+  private final String sourceVertexLabel;
+
+  /**
+   * The edge direction to consider.
+   */
+  private final Neighborhood.EdgeDirection edgeDirection;
+
+  /**
+   * The label of the neighboring vertices that should be connected.
+   */
+  private final String neighborhoodVertexLabel;
+
+  /**
+   * The label of the created edge between the neighbors.
+   */
+  private final String newEdgeLabel;
+
+  /**
+   * The constructor to connect the neighbors of vertices with a certain label.
+   *
+   * @param sourceVertexLabel The label of the vertices the neighborhood is connected.
+   * @param edgeDirection The edge direction to consider.
+   * @param neighborhoodVertexLabel The label of the neighboring vertices that should be connected.
+   * @param newEdgeLabel The label of the created edge between the neighbors.
+   */
+  public ConnectNeighbors(String sourceVertexLabel, Neighborhood.EdgeDirection edgeDirection,
+                          String neighborhoodVertexLabel, String newEdgeLabel) {
+    this.sourceVertexLabel = sourceVertexLabel;
+    this.edgeDirection = edgeDirection;
+    this.neighborhoodVertexLabel = neighborhoodVertexLabel;
+    this.newEdgeLabel = newEdgeLabel;
+  }
+
+
+  @Override
+  public LogicalGraph execute(LogicalGraph graph) {
+
+    // determine the vertices the neighborhood should be calculated for
+    DataSet<Vertex> verticesByLabel = graph.getVerticesByLabel(sourceVertexLabel);
+
+    // prepare the graph
+    LogicalGraph reducedGraph = graph
+        .vertexInducedSubgraph(new LabelIsIn<>(sourceVertexLabel, neighborhoodVertexLabel));
+
+    // determine the neighborhood by edge direction
+    DataSet<Tuple2<Vertex, List<Neighborhood.VertexPojo>>> neighborhood =
+        Neighborhood.getPerVertex(reducedGraph, verticesByLabel, edgeDirection);
+
+    // calculate the new edges and add them to the original graph
+    if (neighborhood != null) {
+      DataSet<Edge> newEdges = neighborhood.flatMap(
+        new CreateCartesianNeighborhoodEdges(graph.getConfig().getEdgeFactory(), newEdgeLabel));
+
+      return graph.getConfig().getLogicalGraphFactory()
+          .fromDataSets(
+              graph.getVertices(),
+              graph.getEdges().union(newEdges));
+    }
+    return null;
```

Comment:
Returning null is not a good pattern. If there are no neighbors that matches this pattern, return the input graph.